tests/conftest.py

```diff
@@ -347,23 +349,16 @@ def validate(self,
                     # Assume ``expected_string`` is a sequence of strings.
                     assert all(string in self.text for string in expected_string)
 
-            if expected_errors is not None:
-                vld = HTMLValidator()
-                done = False
-                count = 3
-                while not done and count > 0:
-                    try:
-                        vld.validate_fragment(self.text)
-                        if len(vld.errors) != expected_errors:
-                            print('Errors for {}: {}'.format(url, len(vld.errors)))
-                            pprint(vld.errors)
-                            assert False
-                        if vld.warnings:
-                            print('Warnings for {}: {}'.format(url, len(vld.warnings)))
-                            pprint(vld.warnings)
-                        done = True
-                    except HTTPError as e:
-                        count -= 1
+            if expected_errors is not None and W3_VALIDATE:
+                # Redo this section using html5validate command line
+                vld = Validator(errors_only=True)
+                tmpname = 'tmphtml.html'
```

Comment:
I would recommend using [tmp_path](http://doc.pytest.org/en/latest/tmpdir.html) to create a directory in which to store `tmphtml.html`. I'd rather not have temp files in the local path, even if the code removes them. (Note that pytest will eventually delete these directories, per https://github.com/pytest-dev/pytest/issues/1120).

Reply:
I didn't see it as any different than what the other validator did with leaving files in the web2py root folder.  If they are temporary and pytest cleans them up then why add the extra code to put them somewhere else??

Reply:
Although now that I look at the docs, it seems that tmp_path / 'tmpfile.html'   Is probably a net zero change.

Reply:
but no idea how you incorporate a fixture into code inside a class in conftest.py

Reply:
I agree that my earlier code putting files in the web2py root folder is sloppy and should also be changed.

To incorporate this, my sketch of the code is:

```
# Create a client for accessing the Runestone server.
class _TestClient(WebClient):
    def __init__(self, web2py_server, web2py_server_address, runestone_name, tmp_path <-- added this):
        self.web2py_server = web2py_server
        self.web2py_server_address = web2py_server_address
        self.tmp_path = tmp_path # <-- added this
```
Then modify the fixture just after it to take `tmp_path` and pass it to the constructor.